### DISCUSSION TITLE
unwanted electric indentation with Eglot/Clangd

### DISCUSSION THREAD

**Discussion by cmm** [joaotavora/eglot#1351] [Q&A] [answered]
Lately I've started experiencing unwanted/unhelpful "electric indentation" (as in, a newline reformats the _previous_ C/C++ syntactic unit in addition to inserting a newline and moving rightwards inside it) in cc-mode with Eglot & Clangd.

`electric-indent-mode` is (very explicitly) off.  The problem only happens with an active Eglot connection (i.e. it disappears if I do `M-x eglot-shutdown`).  This seems to suggest that Eglot somehow involves the language server in indentation (or doesn't prevent it from being involved?), or perhaps Eglot tweaks some Emacs bits that are relevant to indentation but not obvious to identify (at least my grepping the Eglot source for "newline", "indent" and "format" has not been terribly heplful).

I'm using the vague "lately" because unfortunately I happen to have upgraded Eglot (to 1.16), Clangd (to 16.0.6), and possibly other Emacs packages at roughly the same time, so no idea where to even start looking for the culprit.  Any ideas welcome!

**Comment by joaotavora** [comment, marked as answer]
My crystal ball tells me it's the `onTypeFormatting` capability of this server.    Use `eglot-ignore-server-capabilities` to mark this capability as unwanted.

**Comment by cmm** [comment]
adding `:documentOnTypeFormattingProvider` to `eglot-ignored-server-capabilities` appears to have done it, please thank your crystal ball on my behalf! :)

**Reply by joaotavora** [reply]
I'm always on the fence whether to automatically leave this feature there or not.  Maybe it wouldn't be a bad idea to make it conditional on electric-indent-mode.  But then again, this "formatting" is not _just_ about indentation, it's about formatting in general.

**Reply by cmm** [reply]
I certainly can't see any kind of over-eager electric indentation/formatting (whether of Emacs or external origin) ever being desirable when working on a legacy codebase (which is most of codebases), but my imagination is limited :man_shrugging:

**Reply by joaotavora** [reply]
In clangd, when you have a suitable clang-format file and someone has actually run it over the full project, it can be useful.  You'll save yourself some review comments if you forget to run that.

And clang-format is pretty powerful and makes some good decisions in incredibly complex C++ situations.  Even if the decision isn't good, it's _a_ decision, i.e .something you don't need to worry about.

So that's why I'm on the fence.

**Reply by joaotavora** [reply]
Also because the longstanding policy of Eglot has been to enable everything it supports by default.  It's a simple policy that avoids the personal preference wars.

Of course, as a user, I also don't like this particular one, and it bites me whenever I have to Emacs -Q
